eth/transactor.go

```diff
@@ -227,16 +205,116 @@ func (t *Transactor) transact(
+	if hasLegacyFlags && has1559Flags {
+		return ErrConflictingGasFlags
+	}
+	// Check if chain supports EIP-1559
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
+	defer cancel()
+	head, err := client.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to call HeaderByNumber: %w", err)
+	}
+	send1559Tx := false
+	if head.BaseFee != nil && !hasLegacyFlags {
+		send1559Tx = true
+		err = determine1559GasPrice(ctx, signer, txopts, client, head)
+		if err != nil {
+			return fmt.Errorf("failed to determine EIP-1559 gas price: %w", err)
+		}
+	} else {
+		// Legacy pricing
+		err = determineLegacyGasPrice(ctx, signer, txopts, client)
+		if err != nil {
+			return fmt.Errorf("failed to determine legacy gas price: %w", err)
+		}
+	}
+
+	// 2. Determine gas limit
+	if txopts.gasLimit > 0 {
+		// Use the specified limit if set
+		signer.GasLimit = txopts.gasLimit
+		return nil
+	} else if txopts.addGasEstimateRatio > 0.0 {
+		// 2.1. Estimate gas
+		signer.NoSend = true
+		dryTx, err := method(client, signer)
+		if err != nil {
+			return fmt.Errorf("tx dry-run err: %w", err)
+		}
+		signer.NoSend = false
+		var typesMsg types.Message
+		londonSigner := types.NewLondonSigner(t.chainId)
+		if send1559Tx {
+			typesMsg, err = dryTx.AsMessage(londonSigner, head.BaseFee)
+		} else {
+			typesMsg, err = dryTx.AsMessage(londonSigner, nil)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to get typesMsg: %w", err)
+		}
+		callMsg := ethereum.CallMsg{
+			From:     typesMsg.From(),
+			To:       typesMsg.To(),
+			GasPrice: typesMsg.GasPrice(),
+			Value:    typesMsg.Value(),
+			Data:     typesMsg.Data(),
+		}
+		estimatedGas, err := client.EstimateGas(ctx, callMsg)
+		if err != nil {
+			return fmt.Errorf("failed to call EstimateGas: %w", err)
+		}
+		// 2.2. Multiply gas limit by the configured ratio
+		signer.GasLimit = uint64(float64(estimatedGas) * (1 + txopts.addGasEstimateRatio))
+	}
+	// If addGasEstimateRatio not specified, just defer to go-ethereum for gas limit estimation
+	return nil
+}
+
+// determine1559GasPrice sets the gas price on the signer based on the EIP-1559 fee model
+func determine1559GasPrice(
+	ctx context.Context, signer *bind.TransactOpts, txopts txOptions, client *ethclient.Client, head *types.Header) error {
+	if txopts.maxPriorityFeePerGasGwei == 0 {
+		gasTipCap, suggestGasTipCapErr := client.SuggestGasTipCap(ctx)
+		if suggestGasTipCapErr != nil {
+			return fmt.Errorf("failed to call SuggestGasTipCap: %w", suggestGasTipCapErr)
+		}
+		signer.GasTipCap = gasTipCap
+	} else {
+		signer.GasTipCap = new(big.Int).SetUint64(txopts.maxPriorityFeePerGasGwei * 1e9)
+	}
+	if txopts.maxFeePerGasGwei == 0 {
+		// Use (maxPriorityFeePerGas + 2x the curent basefee), the same heuristic as go-ethereum
+		gasFeeCap := new(big.Int).Add(
+			signer.GasTipCap,
+			new(big.Int).Mul(head.BaseFee, big.NewInt(2)),
+		)
+		signer.GasFeeCap = gasFeeCap
```

Comment:
why do you need gasFeeCap variable? signer.GasFeeCap = new(big.Int).Add(...) works the same

Reply:
Removed.